File: app/src/main/java/com/zack/enderplan/view/widget/ItemView.java
package com.zack.enderplan.view.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.zack.enderplan.R;

public class ItemView extends FrameLayout {

    private Drawable mIconDrawable;
    private String mTitleStr;
    private CharSequence mDscptStr;

    private TextView mDscptText;

    public ItemView(Context context) {
        super(context);
        init(null, 0);
    }

    public ItemView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public ItemView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        loadAttrs(attrs, defStyle);
        initViews();
    }

    private void loadAttrs(AttributeSet attrs, int defStyle) {
        TypedArray ta = getContext().obtainStyledAttributes(attrs, R.styleable.ItemView, defStyle, 0);
        mIconDrawable = ta.getDrawable(R.styleable.ItemView_iconImageSrc);
        mTitleStr = ta.getString(R.styleable.ItemView_titleText);
        mDscptStr = ta.getString(R.styleable.ItemView_descriptionText);
        ta.recycle();
    }

    private void initViews() {
        inflate(getContext(), R.layout.widget_item_view, this);

        ((ImageView) findViewById(R.id.image_icon)).setImageDrawable(mIconDrawable);
        ((TextView) findViewById(R.id.text_title)).setText(mTitleStr);

        mDscptText = (TextView) findViewById(R.id.text_dscpt);
        setDescriptionText(mDscptStr);
    }

    public void setDescriptionText(CharSequence text) {
        mDscptText.setText(text);
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/widget/ItemView.kt
package net.zackzhang.app.end.view.widget

import android.content.Context
import android.content.res.ColorStateList
import android.util.AttributeSet
import android.widget.FrameLayout
import kotlinx.android.synthetic.main.widget_item_view.view.*
import net.zackzhang.app.end.R
import net.zackzhang.app.end.util.StringUtil

class ItemView : FrameLayout {

    private var mDscptTextColor = 0

    constructor(context: Context) : super(context) {
        init(null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        inflate(context, R.layout.widget_item_view, this)

        mDscptTextColor = text_dscpt.currentTextColor

        val ta = context.obtainStyledAttributes(attrs, R.styleable.ItemView, defStyle, 0)
        image_icon.setImageDrawable(ta.getDrawable(R.styleable.ItemView_iconImageSrc))
        text_title.text = ta.getString(R.styleable.ItemView_titleText)
        setDescriptionText(ta.getString(R.styleable.ItemView_descriptionText))
        ta.recycle()
    }

    fun setDescriptionText(text: CharSequence?) {
        text_dscpt.text = colorDscptText(text)
    }

    fun setThemeColor(color: Int) {
        setImageIconTintColor(color)
        setDscptTextColor(color)
    }

    fun setImageIconTintColor(color: Int) {
        image_icon.imageTintList = ColorStateList.valueOf(color)
    }

    fun setDscptTextColor(color: Int) {
        if (mDscptTextColor == color) return
        mDscptTextColor = color
        text_dscpt.text = colorDscptText(text_dscpt.text)
    }

    private fun colorDscptText(text: CharSequence?) =
            if (text == null) null else StringUtil.addSpan(text, StringUtil.SPAN_COLOR, mDscptTextColor)
}
